packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -267,13 +264,20 @@ export default createRule<Options, MessageId>({
       if (isStrictCompilerOptionEnabled(compilerOptions, 'strictNullChecks')) {
         const UNDEFINED = ts.TypeFlags.Undefined;
         const NULL = ts.TypeFlags.Null;
+
+        const NULLISH =
+          node.operator === '==' || node.operator === '!='
+            ? NULL | UNDEFINED
+            : NULL;
+
         if (
           (leftType.flags === UNDEFINED &&
-            !typeContainsFlag(rightType, UNDEFINED)) ||
+            !isTypeFlagSet(rightType, UNDEFINED, true)) ||
```

Comment:
It's niche, but we might want similar "NULLISH" handling for the `== undefined` case, otherwise the error will false positive on code like:

```ts
function foo(x: string | null) {
    if(x == undefined) {}
}
```